src/__private_api.rs

```diff
@@ -0,0 +1,321 @@
+    StaticLevelWarn => Warn,
+    StaticLevelInfo => Info,
+    StaticLevelDebug => Debug,
+    StaticLevelTrace => Trace,
+];
+
+// `LogArgs`.
+
+impl LogArgs for &str {
+    #[inline]
+    fn with(self, f: impl FnOnce(Arguments)) {
+        f(format_args!("{self}"))
+    }
+}
+
+impl LogArgs for Arguments<'_> {
+    fn with(self, f: impl FnOnce(Arguments)) {
+        f(self)
+    }
+}
+
+// `LogTarget`.
+
+impl LogTarget for &str {
+    #[inline]
+    fn with(self, _module_path: &'static str, f: impl FnOnce(&str)) {
+        f(self)
+    }
+}
+
+#[derive(Debug)]
+pub struct TargetIsModulePath;
+
+impl LogTarget for TargetIsModulePath {
+    #[inline]
+    fn with(self, module_path: &'static str, f: impl FnOnce(&str)) {
+        f(module_path)
+    }
+}
+
+// `LogKVs`.
+
+impl LogKVs for &[(&str, &LogKvValue<'_>)] {
+    #[inline]
+    fn with(self, f: impl FnOnce(&[(&str, &LogKvValue)])) {
+        f(self)
+    }
+}
+
+#[derive(Debug)]
+pub struct EmptyKVs;
+
+impl LogKVs for EmptyKVs {
+    #[inline]
+    fn with(self, f: impl FnOnce(&[(&str, &LogKvValue)])) {
+        f(&[])
+    }
+}
+
+// Log functions.
+
+fn log_0(
+    &(module_path, file): &'static (&'static str, &'static str),
+    line: u32,
+    level: Level,
+    args: Arguments,
+    target: &str,
+    kvs: &[(&str, &LogKvValue)],
+) {
+    #[cfg(not(feature = "kv_unstable"))]
+    if !kvs.is_empty() {
+        panic!(
+            "key-value support is experimental and must be enabled using the `kv_unstable` feature"
+        );
+    }
+
+    let mut builder = Record::builder();
+
+    builder
+        .args(args)
+        .level(level)
+        .target(target)
+        .module_path_static(Some(module_path))
+        .file_static(Some(file))
+        .line(Some(line));
+
+    #[cfg(feature = "kv_unstable")]
+    builder.key_values(&kvs);
+
+    crate::logger().log(&builder.build());
+}
+
+fn log_1<K>(
+    module_path_and_file: &'static (&'static str, &'static str),
+    line: u32,
+    level: Level,
+    args: Arguments,
+    target: &str,
+    kvs: K,
+) where
+    K: LogKVs,
+{
+    kvs.with(|kvs| log_0(module_path_and_file, line, level, args, target, kvs));
+}
+
+fn log_2<T, K>(
+    module_path_and_file: &'static (&'static str, &'static str),
+    line: u32,
+    level: Level,
+    args: Arguments,
+    target: T,
+    kvs: K,
+) where
+    T: LogTarget,
+    K: LogKVs,
+{
+    target.with(module_path_and_file.0, |target| {
+        log_1(module_path_and_file, line, level, args, target, kvs)
+    });
+}
+
+pub fn log_3<A, T, K>(
+    module_path_and_file: &'static (&'static str, &'static str),
+    line: u32,
+    level: Level,
+    args: A,
+    target: T,
+    kvs: K,
+) where
+    A: LogArgs,
+    T: LogTarget,
+    K: LogKVs,
+{
+    args.with(|args| log_2(module_path_and_file, line, level, args, target, kvs))
+}
+
+pub fn log<L, A, T, K>(
+    module_path_and_file: &'static (&'static str, &'static str),
+    line: u32,
+    level: L,
+    args: A,
```

Comment:
I think you could try passing `&Arguments<'_>` here and in all `log_*` instead of passing generics here.

I still think losing `Arguments::as_str` as an optimization is bad and passing it by reference could make up for it.

Reply:
I have added a const function to check whether the format argument is a string literal as a not perfect replacement for `Arguments::as_str`, that’s why I used a generic argument here. Sometimes I need to pass `&str` as `args`.

As for `&Arguments`, I think I should do some test, then decide which one to use.

Reply:
> I have added a const function to check whether the format argument is a string literal as a not perfect replacement for `Arguments::as_str`

I was actually referring to using `Argument::as_str` in `Log::log`, the implementation might be able to avoid heap alloc using `Argument::as_str`, passing `&str` here breaks that optimization.

Reply:
I am not sure I understand. `Log::log` is not implemented by this crate, how do I utilize `Argument::as_str`?

Reply:
> I am not sure I understand. `Log::log` is not implemented by this crate, so how do I utilize `Argument::as_str`?

Implementator of `Log::log` could use `Argument::as_str` to avoid heap allocation.

Now that you decided to pass arg as `&str` if possible, they can no longer utilize this optimization.

Reply:
I see your point. Maybe this can be optimized by `format_args!` as a special case?

Reply:
> I see your point. Maybe this can be optimized by `format_args!` as a special case?

Ideally yes, but that will require the args to be `&'static str`, which involves lifetime and is a bit hard to do.

I guess we can ask @m-ou-se for this?

Reply:
A more simple could be `impl From<&'static str> for Arguments<'_>`, utilizing [`Arguments::new_const`](https://doc.rust-lang.org/stable/src/core/fmt/mod.rs.html#297).